realhf/impl/model/nn/flatten_param.py

```diff
@@ -78,22 +70,41 @@ def slice_intervals(
     output_size: int,
 ) -> torch.Tensor:
     assert len(tensor.shape) == 1
-    if len(intervals_cpu) == 1:
-        return tensor[intervals_cpu[0][0] : intervals_cpu[0][1]]
-    elif len(intervals_cpu) <= MAX_PYTORCH_N_INTERVALS:
-        return torch.cat([tensor[start:end] for start, end in intervals_cpu])
-
-    interval_sizes = intervals[:, 1] - intervals[:, 0]
-    offsets = torch.nn.functional.pad(interval_sizes.cumsum(0)[:-1], (1, 0), value=0)
-    assert tensor.dtype == torch.half
-    return interval_op_cuda.slice_intervals_cuda_half(
-        tensor,
-        intervals,
-        interval_sizes,
-        offsets,
-        max_interval_size,
-        output_size,
+    assert sum(intervals[:, 1] - intervals[:, 0]) == output_size
+    assert torch.all((intervals[:, 1] - intervals[:, 0]) <= max_interval_size)
+    assert intervals[:, 0].min().item() >= 0, (
+        intervals[:, 0].min().item(),
+        tensor.shape[0],
     )
+    assert intervals[:, 1].max().item() <= tensor.shape[0], (
+        tensor.shape[0],
+        intervals[:, 1].max().item(),
+    )
+
+    # NOTE: these following assertions will take ten years to finish.
+    # assert len(intervals) == len(intervals_cpu)
+    # assert len(set([x[0] for x in intervals_cpu])) == len(intervals_cpu)
+
+    res = torch._C._nn.flatten_dense_tensors(
+        [tensor[start:end] for start, end in intervals_cpu]
+    )
+    return res
+    # if len(intervals_cpu) == 1:
+    #     return tensor[intervals_cpu[0][0] : intervals_cpu[0][1]]
+    # elif len(intervals_cpu) <= MAX_PYTORCH_N_INTERVALS:
+    #     return torch.cat([tensor[start:end] for start, end in intervals_cpu])
+
+    # interval_sizes = intervals[:, 1] - intervals[:, 0]
+    # offsets = torch.nn.functional.pad(interval_sizes.cumsum(0)[:-1], (1, 0), value=0)
+    # assert tensor.dtype == torch.half
+    # return interval_op_cuda.slice_intervals_cuda_half(
+    #     tensor,
+    #     intervals,
+    #     interval_sizes,
+    #     offsets,
+    #     max_interval_size,
+    #     output_size,
+    # )
```

Comment:
Maybe add a TODO to explain commented code here.

Reply:
Added FIXME.